File: backend/controllers/adminController.js
const User = require('../models/User');
const { Subject, Topic } = require('../models/Subject');
const Question = require('../models/Question');
const { Exam } = require('../models/Exam');
const Result = require('../models/Result');
const db = require('../config/database');

exports.getDashboardStats = async (req, res) => {
    try {
        const [
            totalUsers,
            totalStudents,
            totalOperators,
            totalSubjects,
            totalQuestions,
            totalExams,
            recentResults
        ] = await Promise.all([
            User.count(),
            User.count({ role: 'student' }),
            User.count({ role: 'operator' }),
            Subject.count(),
            Question.count(),
            Exam.count(),
            Result.findAll({ limit: 10 })
        ]);

        const examStats = totalExams > 0 ? await Promise.all(
            (await Exam.findAll({ limit: totalExams })).map(async exam => {
                const stats = await Result.getStatistics(exam.id);
                return { examId: exam.id, examName: exam.exam_name, ...stats };
            })
        ) : [];

        res.json({
            success: true,
            data: {
                totalUsers,
                totalStudents,
                totalOperators,
                totalSubjects,
                totalQuestions,
                totalExams,
                recentResults,
                examStats
            }
        });
    } catch (error) {
        console.error('Dashboard stats error:', error);
        res.status(500).json({
            success: false,
            message: 'Failed to get dashboard stats'
        });
    }
};

exports.getAllUsers = async (req, res) => {
    try {
        const { role, search, page = 1, limit = 20 } = req.query;
        const offset = (page - 1) * limit;

        const users = await User.findAll({
            role,
            search,
            limit: parseInt(limit),
            offset
        });

        const total = await User.count({ role });

        res.json({
            success: true,
            data: {
                users: users.map(u => ({
                    id: u.id,
                    userId: u.user_id,
                    username: u.username,
                    email: u.email,
                    role: u.role,
                    firstName: u.first_name,
                    lastName: u.last_name,
                    phone: u.phone,
                    department: u.department,
                    isActive: u.is_active,
                    lastLogin: u.last_login,
                    createdAt: u.created_at
                })),
                pagination: {
                    page: parseInt(page),
                    limit: parseInt(limit),
                    total,
                    pages: Math.ceil(total / limit)
                }
            }
        });
    } catch (error) {
        console.error('Get users error:', error);
        res.status(500).json({
            success: false,
            message: 'Failed to get users'
        });
    }
};

exports.createUser = async (req, res) => {
    try {
        const { username, email, password, role, firstName, lastName, phone, department } = req.body;

        const existingUsername = await User.findByUsername(username);
        if (existingUsername) {
            return res.status(400).json({
                success: false,
                message: 'Username already exists'
            });
        }

        const existingEmail = await User.findByEmail(email);
        if (existingEmail) {
            return res.status(400).json({
                success: false,
                message: 'Email already registered'
            });
        }

        const user = await User.create({
            username,
            email,
            password,
            role,
            firstName,
            lastName,
            phone,
            department
        });

        await db.query(
            'INSERT INTO activity_logs (user_id, action, module, description) VALUES (?, ?, ?, ?)',
            [req.user.id, 'CREATE_USER', 'User Management', `Created user: ${username} with role: ${role}`]
        );

        res.status(201).json({
            success: true,
            message: 'User created successfully',
            data: { userId: user.userId, username, email, role }
        });
    } catch (error) {
        console.error('Create user error:', error);
        res.status(500).json({
            success: false,
            message: 'Failed to create user'
        });
    }
};

exports.updateUser = async (req, res) => {
    try {
        const { id } = req.params;
        const { firstName, lastName, phone, department, isActive } = req.body;

        const user = await User.findById(id);
        if (!user) {
            return res.status(404).json({
                success: false,
                message: 'User not found'
            });
        }

        if (user.role === 'admin' && req.user.role !== 'admin') {
            return res.status(403).json({
                success: false,
                message: 'Cannot modify admin user'
            });
        }

        await User.update(id, { firstName, lastName, phone, department, isActive });

        await db.query(
            'INSERT INTO activity_logs (user_id, action, module, description) VALUES (?, ?, ?, ?)',
            [req.user.id, 'UPDATE_USER', 'User Management', `Updated user: ${user.username}`]
        );

        res.json({
            success: true,
            message: 'User updated successfully'
        });
    } catch (error) {
        console.error('Update user error:', error);
        res.status(500).json({
            success: false,
            message: 'Failed to update user'
        });
    }
};

exports.deleteUser = async (req, res) => {
    try {
        const { id } = req.params;

        const user = await User.findById(id);
        if (!user) {
            return res.status(404).json({
                success: false,
                message: 'User not found'
            });
        }

        if (user.role === 'admin') {
            return res.status(403).json({
                success: false,
                message: 'Cannot delete admin user'
            });
        }

        if (user.id === req.user.id) {
            return res.status(400).json({
                success: false,
                message: 'Cannot delete your own account'
            });
        }

        await User.delete(id);

        await db.query(
            'INSERT INTO activity_logs (user_id, action, module, description) VALUES (?, ?, ?, ?)',
            [req.user.id, 'DELETE_USER', 'User Management', `Deleted user: ${user.username}`]
        );

        res.json({
            success: true,
            message: 'User deleted successfully'
        });
    } catch (error) {
        console.error('Delete user error:', error);
        res.status(500).json({
            success: false,
            message: 'Failed to delete user'
        });
    }
};

exports.getActivityLogs = async (req, res) => {
    try {
        const { page = 1, limit = 50 } = req.query;
        const offset = (page - 1) * limit;

        const logs = await db.query(`
            SELECT al.*, u.username 
            FROM activity_logs al 
            LEFT JOIN users u ON al.user_id = u.id 
            ORDER BY al.created_at DESC 
            LIMIT ? OFFSET ?
        `, [parseInt(limit), offset]);

        const [{ total }] = await db.query('SELECT COUNT(*) as total FROM activity_logs');

        res.json({
            success: true,
            data: {
                logs,
                pagination: {
                    page: parseInt(page),
                    limit: parseInt(limit),
                    total,
                    pages: Math.ceil(total / limit)
                }
            }
        });
    } catch (error) {
        console.error('Get activity logs error:', error);
        res.status(500).json({
            success: false,
            message: 'Failed to get activity logs'
        });
    }
};

exports.getExamPolicies = async (req, res) => {
    try {
        const policies = await db.query('SELECT * FROM exam_policies WHERE is_active = 1');
        
        const policyMap = {};
        policies.forEach(p => {
            policyMap[p.policy_key] = p.policy_value;
        });

        res.json({
            success: true,
            data: policyMap
        });
    } catch (error) {
        console.error('Get policies error:', error);
        res.status(500).json({
            success: false,
            message: 'Failed to get policies'
        });
    }
};

exports.updateExamPolicy = async (req, res) => {
    try {
        const { policyKey, policyValue, policyName, description } = req.body;

        const [existing] = await db.query('SELECT * FROM exam_policies WHERE policy_key = ?', [policyKey]);

        if (existing) {
            await db.query(
                'UPDATE exam_policies SET policy_value = ?, updated_by = ? WHERE policy_key = ?',
                [policyValue, req.user.id, policyKey]
            );
        } else {
            await db.query(
                'INSERT INTO exam_policies (policy_name, policy_key, policy_value, description, updated_by) VALUES (?, ?, ?, ?, ?)',
                [policyName, policyKey, policyValue, description, req.user.id]
            );
        }

        res.json({
            success: true,
            message: 'Policy updated successfully'
        });
    } catch (error) {
        console.error('Update policy error:', error);
        res.status(500).json({
            success: false,
            message: 'Failed to update policy'
        });
    }
};

File: OLESS/Controllers/AdminController.cs
using System;
using System.Collections.Generic;
using Microsoft.AspNetCore.Mvc;
using Microsoft.AspNetCore.Http;
using OLESS.Models;

namespace OLESS.Controllers
{
    public class AdminController : Controller
    {
        private readonly DatabaseContext _db;

        public AdminController(DatabaseContext db)
        {
            _db = db;
        }

        public IActionResult Index()
        {
            if (!IsAuthenticated() || GetUserRole() != "Admin")
            {
                return RedirectToAction("Login", "Account");
            }

            var userRepo = new UserRepository(_db);
            var subjectRepo = new SubjectRepository(_db);
            var questionRepo = new QuestionRepository(_db);
            var examineeRepo = new ExamineeRepository(_db);
            var resultRepo = new ResultRepository(_db);

            ViewBag.TotalUsers = userRepo.GetAll().Count;
            ViewBag.TotalStudents = userRepo.GetByRole("Student").Count;
            ViewBag.TotalSubjects = subjectRepo.GetActive().Count;
            ViewBag.TotalQuestions = questionRepo.GetAll().Count;
            ViewBag.RecentResults = resultRepo.GetAll().GetRange(0, Math.Min(10, resultRepo.GetAll().Count));

            return View();
        }

        public IActionResult Dashboard()
        {
            return Index();
        }

        public IActionResult ManageUsers()
        {
            if (!IsAuthenticated() || GetUserRole() != "Admin")
            {
                return RedirectToAction("Login", "Account");
            }

            var userRepo = new UserRepository(_db);
            ViewBag.Users = userRepo.GetAll();
            ViewBag.Roles = new List<string> { "Admin", "Student", "Operator", "Controller" };

            return View();
        }

        [HttpPost]
        public IActionResult CreateUser(string username, string email, string password, string role, string firstName, string lastName)
        {
            if (!IsAuthenticated() || GetUserRole() != "Admin")
            {
                return RedirectToAction("Login", "Account");
            }

            try
            {
                var userRepo = new UserRepository(_db);

                if (userRepo.GetByUsername(username) != null)
                {
                    TempData["Error"] = "Username already exists.";
                    return RedirectToAction("ManageUsers");
                }

                var user = new User
                {
                    Username = username,
                    PasswordHash = password,
                    Email = email,
                    Role = role,
                    FirstName = firstName,
                    LastName = lastName
                };

                int userId = userRepo.Create(user);

                if (role == "Student")
                {
                    var examineeRepo = new ExamineeRepository(_db);
                    var examinee = new Examinee
                    {
                        UserID = userId,
                        EnrollmentNo = "ENR" + DateTime.Now.ToString("yyyyMMdd") + userId.ToString().PadLeft(4, '0'),
                        DateOfBirth = DateTime.Now.AddYears(-18)
                    };
                    examineeRepo.Create(examinee);
                }

                TempData["Success"] = "User created successfully!";
            }
            catch (Exception ex)
            {
                TempData["Error"] = "Error creating user: " + ex.Message;
            }

            return RedirectToAction("ManageUsers");
        }

        [HttpPost]
        public IActionResult UpdateUser(int userId, string email, string firstName, string lastName, bool isActive)
        {
            if (!IsAuthenticated() || GetUserRole() != "Admin")
            {
                return RedirectToAction("Login", "Account");
            }

            try
            {
                var userRepo = new UserRepository(_db);
                var user = userRepo.GetById(userId);
                if (user != null)
                {
                    user.Email = email;
                    user.FirstName = firstName;
                    user.LastName = lastName;
                    user.IsActive = isActive;
                    userRepo.Update(user);
                    TempData["Success"] = "User updated successfully!";
                }
            }
            catch (Exception ex)
            {
                TempData["Error"] = "Error updating user: " + ex.Message;
            }

            return RedirectToAction("ManageUsers");
        }

        [HttpPost]
        public IActionResult DeleteUser(int userId)
        {
            if (!IsAuthenticated() || GetUserRole() != "Admin")
            {
                return RedirectToAction("Login", "Account");
            }

            int currentUserId = HttpContext.Session.GetInt32("UserID") ?? 0;
            if (userId == currentUserId)
            {
                TempData["Error"] = "You cannot delete your own account.";
                return RedirectToAction("ManageUsers");
            }

            try
            {
                var parameters = new[] { new Microsoft.Data.SqlClient.SqlParameter("@UserID", userId) };
                _db.ExecuteNonQuery("UPDATE Users SET IsActive = 0 WHERE UserID = @UserID", parameters);
                TempData["Success"] = "User deactivated successfully!";
            }
            catch (Exception ex)
            {
                TempData["Error"] = "Error deleting user: " + ex.Message;
            }

            return RedirectToAction("ManageUsers");
        }

        public IActionResult ManageSubjects()
        {
            if (!IsAuthenticated() || GetUserRole() != "Admin")
            {
                return RedirectToAction("Login", "Account");
            }

            var subjectRepo = new SubjectRepository(_db);
            ViewBag.Subjects = subjectRepo.GetAll();

            return View();
        }

        [HttpPost]
        public IActionResult CreateSubject(string subjectCode, string subjectName, string description, int credits)
        {
            if (!IsAuthenticated() || GetUserRole() != "Admin")
            {
                return RedirectToAction("Login", "Account");
            }

            try
            {
                int userId = HttpContext.Session.GetInt32("UserID") ?? 0;
                var subjectRepo = new SubjectRepository(_db);

                var subject = new Subject
                {
                    SubjectCode = subjectCode,
                    SubjectName = subjectName,
                    Description = description,
                    Credits = credits,
                    CreatedBy = userId
                };

                subjectRepo.Create(subject);
                TempData["Success"] = "Subject created successfully!";
            }
            catch (Exception ex)
            {
                TempData["Error"] = "Error creating subject: " + ex.Message;
            }

            return RedirectToAction("ManageSubjects");
        }

        [HttpPost]
        public IActionResult UpdateSubject(int subjectId, string subjectCode, string subjectName, string description, int credits, bool isActive)
        {
            if (!IsAuthenticated() || GetUserRole() != "Admin")
            {
                return RedirectToAction("Login", "Account");
            }

            try
            {
                var subjectRepo = new SubjectRepository(_db);
                var subject = subjectRepo.GetById(subjectId);
                if (subject != null)
                {
                    subject.SubjectCode = subjectCode;
                    subject.SubjectName = subjectName;
                    subject.Description = description;
                    subject.Credits = credits;
                    subject.IsActive = isActive;
                    subjectRepo.Update(subject);
                    TempData["Success"] = "Subject updated successfully!";
                }
            }
            catch (Exception ex)
            {
                TempData["Error"] = "Error updating subject: " + ex.Message;
            }

            return RedirectToAction("ManageSubjects");
        }

        [HttpPost]
        public IActionResult DeleteSubject(int subjectId)
        {
            if (!IsAuthenticated() || GetUserRole() != "Admin")
            {
                return RedirectToAction("Login", "Account");
            }

            try
            {
                var subjectRepo = new SubjectRepository(_db);
                subjectRepo.Delete(subjectId);
                TempData["Success"] = "Subject deleted successfully!";
            }
            catch (Exception ex)
            {
                TempData["Error"] = "Error deleting subject: " + ex.Message;
            }

            return RedirectToAction("ManageSubjects");
        }

        public IActionResult ManageTopics()
        {
            if (!IsAuthenticated() || GetUserRole() != "Admin")
            {
                return RedirectToAction("Login", "Account");
            }

            var subjectRepo = new SubjectRepository(_db);
            var topicRepo = new TopicRepository(_db);

            ViewBag.Subjects = subjectRepo.GetActive();
            ViewBag.Topics = new List<Topic>();

            return View();
        }

        public IActionResult GetTopicsBySubject(int subjectId)
        {
            if (!IsAuthenticated() || GetUserRole() != "Admin")
            {
                return Json(new { error = "Unauthorized" });
            }

            var topicRepo = new TopicRepository(_db);
            var topics = topicRepo.GetBySubject(subjectId);
            return Json(topics);
        }

        [HttpPost]
        public IActionResult CreateTopic(int subjectId, string topicName, string description, int weightage)
        {
            if (!IsAuthenticated() || GetUserRole() != "Admin")
            {
                return RedirectToAction("Login", "Account");
            }

            try
            {
                var topicRepo = new TopicRepository(_db);

                var topic = new Topic
                {
                    SubjectID = subjectId,
                    TopicName = topicName,
                    Description = description,
                    Weightage = weightage
                };

                topicRepo.Create(topic);
                TempData["Success"] = "Topic created successfully!";
            }
            catch (Exception ex)
            {
                TempData["Error"] = "Error creating topic: " + ex.Message;
            }

            return RedirectToAction("ManageTopics");
        }

        [HttpPost]
        public IActionResult UpdateTopic(int topicId, string topicName, string description, int weightage, bool isActive)
        {
            if (!IsAuthenticated() || GetUserRole() != "Admin")
            {
                return RedirectToAction("Login", "Account");
            }

            try
            {
                var topicRepo = new TopicRepository(_db);
                var topic = topicRepo.GetById(topicId);
                if (topic != null)
                {
                    topic.TopicName = topicName;
                    topic.Description = description;
                    topic.Weightage = weightage;
                    topic.IsActive = isActive;
                    topicRepo.Update(topic);
                    TempData["Success"] = "Topic updated successfully!";
                }
            }
            catch (Exception ex)
            {
                TempData["Error"] = "Error updating topic: " + ex.Message;
            }

            return RedirectToAction("ManageTopics");
        }

        [HttpPost]
        public IActionResult DeleteTopic(int topicId)
        {
            if (!IsAuthenticated() || GetUserRole() != "Admin")
            {
                return RedirectToAction("Login", "Account");
            }

            try
            {
                var topicRepo = new TopicRepository(_db);
                topicRepo.Delete(topicId);
                TempData["Success"] = "Topic deleted successfully!";
            }
            catch (Exception ex)
            {
                TempData["Error"] = "Error deleting topic: " + ex.Message;
            }

            return RedirectToAction("ManageTopics");
        }

        public IActionResult ManageQuestions()
        {
            if (!IsAuthenticated() || GetUserRole() != "Admin")
            {
                return RedirectToAction("Login", "Account");
            }

            var subjectRepo = new SubjectRepository(_db);
            var topicRepo = new TopicRepository(_db);
            var questionRepo = new QuestionRepository(_db);

            ViewBag.Subjects = subjectRepo.GetActive();
            ViewBag.Topics = topicRepo.GetBySubject(1);
            ViewBag.Questions = questionRepo.GetAll();

            return View();
        }

        [HttpPost]
        public IActionResult CreateQuestion(int subjectId, int? topicId, string questionType, string questionText, string optionA, string optionB, string optionC, string optionD, string correctAnswer, string difficultyLevel, int marks, string explanation)
        {
            if (!IsAuthenticated() || GetUserRole() != "Admin")
            {
                return RedirectToAction("Login", "Account");
            }

            try
            {
                int userId = HttpContext.Session.GetInt32("UserID") ?? 0;
                var questionRepo = new QuestionRepository(_db);

                var question = new Question
                {
                    SubjectID = subjectId,
                    TopicID = topicId,
                    QuestionType = questionType,
                    QuestionText = questionText,
                    OptionA = optionA,
                    OptionB = optionB,
                    OptionC = optionC,
                    OptionD = optionD,
                    CorrectAnswer = correctAnswer,
                    DifficultyLevel = difficultyLevel,
                    Marks = marks,
                    Explanation = explanation,
                    CreatedBy = userId
                };

                questionRepo.Create(question);
                TempData["Success"] = "Question created successfully!";
            }
            catch (Exception ex)
            {
                TempData["Error"] = "Error creating question: " + ex.Message;
            }

            return RedirectToAction("ManageQuestions");
        }

        [HttpPost]
        public IActionResult UpdateQuestion(int questionId, int subjectId, int? topicId, string questionType, string questionText, string optionA, string optionB, string optionC, string optionD, string correctAnswer, string difficultyLevel, int marks, string explanation, bool isActive)
        {
            if (!IsAuthenticated() || GetUserRole() != "Admin")
            {
                return RedirectToAction("Login", "Account");
            }

            try
            {
                var questionRepo = new QuestionRepository(_db);
                var question = questionRepo.GetById(questionId);
                if (question != null)
                {
                    question.SubjectID = subjectId;
                    question.TopicID = topicId;
                    question.QuestionType = questionType;
                    question.QuestionText = questionText;
                    question.OptionA = optionA;
                    question.OptionB = optionB;
                    question.OptionC = optionC;
                    question.OptionD = optionD;
                    question.CorrectAnswer = correctAnswer;
                    question.DifficultyLevel = difficultyLevel;
                    question.Marks = marks;
                    question.Explanation = explanation;
                    question.IsActive = isActive;
                    questionRepo.Update(question);
                    TempData["Success"] = "Question updated successfully!";
                }
            }
            catch (Exception ex)
            {
                TempData["Error"] = "Error updating question: " + ex.Message;
            }

            return RedirectToAction("ManageQuestions");
        }

        [HttpPost]
        public IActionResult DeleteQuestion(int questionId)
        {
            if (!IsAuthenticated() || GetUserRole() != "Admin")
            {
                return RedirectToAction("Login", "Account");
            }

            try
            {
                var questionRepo = new QuestionRepository(_db);
                questionRepo.Delete(questionId);
                TempData["Success"] = "Question deleted successfully!";
            }
            catch (Exception ex)
            {
                TempData["Error"] = "Error deleting question: " + ex.Message;
            }

            return RedirectToAction("ManageQuestions");
        }

        public IActionResult ViewResults()
        {
            if (!IsAuthenticated() || GetUserRole() != "Admin")
            {
                return RedirectToAction("Login", "Account");
            }

            var resultRepo = new ResultRepository(_db);
            var subjectRepo = new SubjectRepository(_db);

            ViewBag.Results = resultRepo.GetAll();
            ViewBag.Subjects = subjectRepo.GetActive();

            return View();
        }

        public IActionResult GetSubjectResults(int subjectId)
        {
            if (!IsAuthenticated() || GetUserRole() != "Admin")
            {
                return Json(new { error = "Unauthorized" });
            }

            var resultRepo = new ResultRepository(_db);
            var results = resultRepo.GetBySubject(subjectId);
            return Json(results);
        }

        public IActionResult GetPassedStudents(int subjectId)
        {
            if (!IsAuthenticated() || GetUserRole() != "Admin")
            {
                return Json(new { error = "Unauthorized" });
            }

            var resultRepo = new ResultRepository(_db);
            var results = resultRepo.GetPassedStudents(subjectId);
            return Json(results);
        }

        public IActionResult GetFailedStudents(int subjectId)
        {
            if (!IsAuthenticated() || GetUserRole() != "Admin")
            {
                return Json(new { error = "Unauthorized" });
            }

            var resultRepo = new ResultRepository(_db);
            var results = resultRepo.GetFailedStudents(subjectId);
            return Json(results);
        }

        public IActionResult Reports()
        {
            if (!IsAuthenticated() || GetUserRole() != "Admin")
            {
                return RedirectToAction("Login", "Account");
            }

            var subjectRepo = new SubjectRepository(_db);
            var resultRepo = new ResultRepository(_db);

            ViewBag.Subjects = subjectRepo.GetActive();
            ViewBag.RecentResults = resultRepo.GetAll();

            return View();
        }

        public IActionResult ExamHistory()
        {
            if (!IsAuthenticated() || GetUserRole() != "Admin")
            {
                return RedirectToAction("Login", "Account");
            }

            var examRepo = new ExamRecordRepository(_db);
            ViewBag.Exams = examRepo.GetAll();

            return View();
        }

        private bool IsAuthenticated()
        {
            return HttpContext.Session.GetInt32("UserID") != null;
        }

        private string GetUserRole()
        {
            return HttpContext.Session.GetString("Role") ?? "";
        }
    }
}
